### ISSUE TITLE
Model outputs

### ISSUE THREAD

**Issue by oseymour** [ultralytics/ultralytics#7052]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I've read the [docs](https://docs.ultralytics.com/modes/predict/#working-with-results) and other issues on here but haven't found anything that has helped me fix my issue. For background, I'm trying to train a model from scratch for a real-time segmentation application. There are only 2 classes and the only output I need are the predicted probabilities of each pixel in the same shape as in the input image. I have a training pipeline already in place and was hoping to be able to drop the PyTorch model into it and train.

I copied the `yolov8-seg.yaml` file and first tried loading the model and predicting with it, just to make sure it would work with my inputs. It ran without error and no detections (expected) but the `probs` field is `None`. So this won't work for me. Here's the code I ran:
``` python
model = YOLO(yolov8n-seg.yaml') # type ultralytics.models.yolo.model.YOLO
input = val_ds[0]['image'].unsqueeze(0).to('cpu') # a 1x3x256x256 Tensor
output = model(input)
```

I went a level deeper and tried:
``` python
model = YOLO('yolov8n-seg.yaml').model # type ultralytics.nn.tasks.SegmentationModel
input = val_ds[0]['image'].unsqueeze(0).to('cpu') # a 1x3x256x256 Tensor
output = model(input)
```
This returned a tuple of length 3, with the following items in each element:
- The first element is a tuple of 3 tensors, 1x66x32x32, 1x66x16x16, 1x66x8x8
- The second element is a single tensor, 1x32x1344
- The final element is another tensor, this one size 1x32x64x64

This seems to be in conflict with the docs which state that `BaseModel.forward()` will return "(torch.Tensor): The output of the network.". I don't see the `forward` method being overridden in `DetectionModel` or `SegmentationModel`. **Are one of these outputs the predicted probabilities? If so, how do I transform them back to image space (so they're the same size as my input image)?**

P.S. I also tried running inference with `YOLO('yolov8n-seg.yaml').model.model # type torch.nn.modules.container.Sequential` but the forward pass failed to run. I don't think I need to go this deep into your abstractions to get the output I want?

Thank you in advance!

**Comment by glenn-jocher** [comment]
@oseymour hello! 👋 For real-time segmentation with only the pixel probabilities needed, you'd ideally use the output from the `SegmentationModel` as it is optimized for such tasks.

In your case, after running the model, you're dealing with multi-scale outputs. These are designed to be used internally for further processing, not directly as pixel probabilities.

To obtain pixel probabilities in the shape of the input image:
1. Use the last output tensor, which is the segmentation map at a reduced scale. 
2. Resize it to match your input image dimensions. The resizing transforms the tensor from the feature space back to the image space.

The actual probability values are given by applying a softmax to the resized segmentation map along the channel dimension. This will give you the probability per pixel for each class.

Concerning the direct use of the sequential model and related problems, it seems you might not need to go so deep. The issue might be due to the specific configuration and preprocessing steps the model expects.

For more details, check out our documentation on working with segmentation models. The docs are comprehensive and can guide you through understanding the output formats and transformations required. If you still face a specific issue, feel free to reach out, and we'll be glad to assist.

Happy segmenting, and thank you for engaging with our community! 🚀 If you have further concerns or need additional clarification, we're here to help.

**Comment by oseymour** [comment]
I managed to get this working by just resizing the SegmentationModel outputs to my input image size. For anybody that finds this issue in the future, be wary of resizing the model outputs during training. Make sure your resizing method is compatible with PyTorch's autograd (or whatever equivalent TensorFlow uses).

I was initially using MONAI's Resize transform but this detaches the output from the computation graph and makes gradient descent impossible. I ended up using torchvision's Resize transform. Works for the spatial dimensions but it not for the channel dimensions.

**Comment by glenn-jocher** [comment]
Great to hear you've resolved the resizing issue, @oseymour! 👍 Indeed, it's crucial to use differentiable operations during training to maintain the computation graph for backpropagation. `torchvision.transforms.Resize` is a good choice for this purpose.

For others reading this: when resizing outputs for segmentation tasks, ensure that the resizing function preserves gradients to allow for effective training. If you encounter similar issues, always check that the operations you're using are compatible with the automatic differentiation mechanisms of your deep learning framework.

Keep up the good work, and if you have any more questions or run into further issues, don't hesitate to ask. Happy coding! 🚀

**Comment by tteresi7** [comment]
@glenn-jocher it seems kind if impractical to expect the end user to do this.  Why can't there be a function added when you retrieve the segmentation from the results, to resize it to the original image size?

**Comment by glenn-jocher** [comment]
@tteresi7 hey there! 🌟 Thanks for your suggestion. You're right; having a convenient function to resize the segmentation output directly to the original image size would indeed make things easier. Here’s a quick snippet you can use in the meantime:

```python
from torchvision.transforms.functional import resize

# Assuming 'output' is your model's segmentation output
resized_output = resize(output, size=(original_height, original_width))
```

We'll consider including a built-in function for future updates to streamline this process. Your feedback helps us improve, so keep it coming! Happy coding! 🚀

**Comment by tteresi7** [comment]
> @tteresi7 hey there! 🌟 Thanks for your suggestion. You're right; having a convenient function to resize the segmentation output directly to the original image size would indeed make things easier. Here’s a quick snippet you can use in the meantime:
> 
> ```python
> from torchvision.transforms.functional import resize
> 
> # Assuming 'output' is your model's segmentation output
> resized_output = resize(output, size=(original_height, original_width))
> ```
> 
> We'll consider including a built-in function for future updates to streamline this process. Your feedback helps us improve, so keep it coming! Happy coding! 🚀

Yeah the problem with this though was that it was padded.  And I had no idea what the padding was.  I actually had to go through LetterBox and sort of hack it to see how much the padding was, because otherwise the resized segmentation was shifted.  Like calling .plot() gave a good result, but resizing the segmentation manually resulted in shifted results due to the padding.

**Comment by glenn-jocher** [comment]
Hey @tteresi7, thanks for pointing that out! 🛠 To handle padding issues when resizing segmentation outputs manually, a good approach is to match the `.plot()` function’s behavior. This involves accounting for the padding introduced during preprocessing. Here’s a simplified way to do it without delving into `LetterBox` directly:

```python
import torch
from torchvision.transforms.functional import resize, pad

# Assuming 'output' is your model's raw segmentation output (1xNUM_CLASSESxHxW)
# And 'original_dimensions' and 'model_input_dimensions' are tuples (height, width)
def resize_and_unpad_segmentation(output, original_dimensions, model_input_dimensions):
    # Resize to model input dimensions first
    resized_output = resize(output, size=model_input_dimensions)
    
    # Calculate padding
    pad_height = (original_dimensions[0] - model_input_dimensions[0]) // 2
    pad_width = (original_dimensions[1] - model_input_dimensions[1]) // 2
    
    # Unpad
    unpadded_output = resized_output[..., pad_height: resized_output.shape[-2]-pad_height, pad_width: resized_output.shape[-1]-pad_width]
    
    return unpadded_output

# Example usage
unpadded_output = resize_and_unpad_segmentation(output, (original_height, original_width), (model_height, model_width))
resized_and_unpadded_output = resize(unpadded_output, size=(original_height, original_width))
```

This method resizes the segmentation output back to the model's input size, then crops out the padding before resizing to the original image size. It should align better with the results from `.plot()`. Let us know if this helps or if you have more insights! Keep experimenting! 🚀

**Comment by tteresi7** [comment]
Sorry to bump this again.  Was there ever an issue or fix for the lack of automatic resizing/padding removal?  It's not really practical to expect the end user to do this, considering standard detection will handle this, but it only does this with segmentation.

**Comment by glenn-jocher** [comment]
Thanks for your feedback! We're actively working on improving segmentation output handling in upcoming releases. For now, you can use `scale_masks()` or `scale_image()` from our ops utilities to handle resizing and padding removal automatically. These functions account for the letterboxing transformations used during preprocessing. For implementation details, see the scale_masks documentation and scale_image documentation. Let us know if you'd like help implementing these in your workflow! 🚀
